File: cpp_optimizer/optimizer/why_fail.cpp
#include "why_fail.h"

void update_failure_reason(BannedSlots& banned, Slot& slot, Grade& only_new_slot_grade, Grade& only_old_slot_grade) {
    int old_total     = only_old_slot_grade.totalScore;
    int new_professor = only_new_slot_grade.professorScore;
    int new_semester  = only_new_slot_grade.semScore;
    int new_capacity  = only_new_slot_grade.capacityScore;
    int new_computer  = only_new_slot_grade.computerScore;
    triplet date = std::make_tuple(slot.week, slot.day, slot.hour);

    if((new_professor + new_semester) <= old_total) {
        banned.ban_dates.insert(date);
    }
    if((new_capacity + new_computer) <= old_total) {
        banned.ban_rooms.insert(slot.room_id);
    }
}

bool is_skippable(BannedSlots& banned, Slot& slot) {
    triplet date = std::make_tuple(slot.week, slot.day, slot.hour);
    return banned.ban_rooms.find(slot.room_id) != banned.ban_rooms.end() ||
           banned.ban_dates.find(date) != banned.ban_dates.end();
}

File: optimizer/why_fail.py
"""
Returns an empty action object.
Used to determine which dates can be skipped (banned).
"""
def init_action():
    return {"ban_dates": set(), "ban_rooms": set()}


"""
Updates action object with knowledge gained from new slot grade and old slot grade.
E.g. if professor + sem grade is worse than the old total grade then we can skip
     that slot in the future because it will have the same problem.
"""
def failure_reason(action, slot, only_new_slot_grade, only_old_slot_grade):
    old_total     = only_old_slot_grade["totalScore"]
    new_professor = only_new_slot_grade["professorScore"]
    new_sem       = only_new_slot_grade["semScore"]
    new_capacity  = only_new_slot_grade["capacityScore"]
    new_computer  = only_new_slot_grade["computerScore"]
    room_id, week, day, hr = slot
    ban_slot = (week, day, hr)

    # <= because values are negative (e.g. -1200 is worse than -1000)
    if new_professor + new_sem <= old_total:
        action["ban_dates"].add(ban_slot)

    if new_capacity + new_computer <= old_total:
        action["ban_rooms"].add(room_id)


"""
Returns True if slot is skippable by the optimizer algorithm.
"""
def is_skippable(slot, action):
    room_id,week,day,hr = slot
    ban_slot = (week, day, hr)
    return ban_slot in action["ban_dates"] or room_id in action["ban_rooms"]

